src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -65,27 +96,133 @@ open class GossipRouter : AbstractRouter() {
-                iwant(receivedFrom, controlMsg.messageIDsList - seenMessages)
-            is Rpc.ControlIWant ->
-                controlMsg.messageIDsList
-                    .mapNotNull { mCache.messages[it] }
-                    .forEach { submitPublishMessage(receivedFrom, it) }
+            is Rpc.ControlGraft -> handleGraft(controlMsg, receivedFrom)
+            is Rpc.ControlPrune -> handlePrune(controlMsg, receivedFrom)
+            is Rpc.ControlIHave -> handleIHave(controlMsg, receivedFrom)
+            is Rpc.ControlIWant -> handleIWant(controlMsg, receivedFrom)
         }
     }
 
+    private fun handleGraft(msg: Rpc.ControlGraft, peer: PeerHandler) {
+        val topic = msg.topicID
+        // ignore GRAFT for unknown topics
+        val meshPeers = mesh[topic] ?: return
+        when {
+            isDirect(peer) ->
+                prune(peer, topic)
+            isBackOff(peer, topic) -> {
+                notifyRouterMisbehavior(peer, 1)
+                if (isBackOffFlood(peer, topic)) {
+                    notifyRouterMisbehavior(peer, 1)
+                }
+                prune(peer, topic)
+            }
+            score.score(peer) < 0 ->
+                prune(peer, topic)
+            meshPeers.size >= coreParams.DHigh && !peer.isOutbound() ->
+                prune(peer, topic)
+            peer !in meshPeers ->
+                graft(peer, topic)
+        }
+    }
+
+    private fun handlePrune(msg: Rpc.ControlPrune, peer: PeerHandler) {
+        val topic = msg.topicID
+        mesh[topic]?.remove(peer)?.also {
+            notifyPruned(peer, topic)
+        }
+        if (this.protocol == PubsubProtocol.Gossip_V_1_1) {
+            if (msg.hasBackoff()) {
+                setBackOff(peer, topic, msg.backoff.seconds.toMillis())
+            } else {
+                setBackOff(peer, topic)
+            }
+            if (score.score(peer) >= scoreParams.acceptPXThreshold) {
+                processPrunePeers(msg.peersList)
+            }
+        } else {
+            if (msg.hasBackoff() || msg.peersCount > 0) {
+                notifyRouterMisbehavior(peer, 1)
+            }
+        }
+    }
+
+    private fun handleIHave(msg: Rpc.ControlIHave, peer: PeerHandler) {
+        val peerScore = score.score(peer)
+        // we ignore IHAVE gossip from any peer whose score is below the gossip threshold
+        if (peerScore < score.params.gossipThreshold) return
+        if (peerIHave.computeIfAbsent(peer) { AtomicInteger() }.incrementAndGet() > params.maxIHaveMessages) {
+            return
+        }
+
+        val iWant = msg.messageIDsList - seenMessages.keys
+        val asked = iAsked.computeIfAbsent(peer) { AtomicInteger() }
+        val maxToAsk = max(0, min(iWant.size, params.maxIHaveLength - asked.get()))
+        asked.addAndGet(iWant.size)
+        iWant(peer, iWant.shuffled(random).subList(0, maxToAsk))
+    }
+
+    private fun handleIWant(msg: Rpc.ControlIWant, peer: PeerHandler) {
```

Comment:
I believe we have to implement this:
> IWANT message responses are limited in the number of retransmissions to a certain peer; in gossipsub v1.0 the router always responds to IWANT messages when the message in the cache. In gossipsub v1.1 the router responds a limited number of times to each peer so that IWANT spam does not cause a signficant drain of resources.

Unless you implemented this somewhere else and I'm missing it.

Reply:
Good catch! 👍 
Missed that restriction

Reply:
Done: https://github.com/libp2p/jvm-libp2p/commit/94a02d0c9dae34d9fdc18ab79a61d99fc9560e49